### ISSUE TITLE
sample recurrent iterations

### ISSUE THREAD

**Issue by hzby** [seal-rg/recurrent-pretraining#15]
Thank you for your excellent work! I’m curious about the motivation behind sampling recurrent iterations during training. Would using fixed recurrent iterations for both training and evaluation potentially lead to better results?

**Comment by JonasGeiping** [comment]
Hi, thank you for your question! If the model is trained with fixed recurrent iterations, then it will not generalize in recurrence steps, so the model will not be able to use fewer steps effectively, and not able to be able to extrapolate to more steps. Often, these models "collapse", or "overthink" when running for too long. Training with the randomized unrolling objective solves this problem.

At evaluation time, we do not evaluate with randomized recurrence, we either set a fixed recurrence, or a per-token stopping criterion.

**Comment by hzby** [comment]
Thank you for your response. I’m still curious—if we don’t aim to generalize in the recurrence steps and instead train with a fixed number of steps, then evaluate using the same fixed number of steps, would this lead to better results? Have you conducted any experiments along these lines?

**Comment by JonasGeiping** [comment]
We have not tested this at large scales. At small scale, the gain from fixing the recurrence is marginal (and the costs high for us, because we are looking for test-time scaling, so extrapolation in steps). 

What you're describing is sometimes refered to more directly as a weight-shared model, and one classic larger(r)-scale example is https://arxiv.org/abs/1909.11942

**Comment by JonasGeiping** [comment]
Closing this for now, feel free to re-open with follow-up questions.

**Comment by shuxiaobo** [comment]
Hi @JonasGeiping ，I have tried to implement the architecture from your paper in Megatron. For controlling the number of recurrences, I experimented with both fixed recurrence steps and a strategy that combines random recurrence with gradient clipping. However, based on the loss, it seems that the fixed recurrence approach performs better. Have you observed a similar phenomenon?

**Comment by shuxiaobo** [comment]
@JonasGeiping 
In terms of loss, the random recurrence approach is about 0.02 higher than the fixed one. The experiments were conducted on a 1B LLaMA model with batch size and sequence length set to [512, 512], which is a small-scale setup. I would appreciate any advice or suggestions you might have.
![Image](https://github.com/user-attachments/assets/c6f4c651-182d-459f-b110-d97a175779cf)

**Comment by JonasGeiping** [comment]
Just be clear, this is fixed recurrence of let's say `r`, evaluated at `r`?

**Comment by shuxiaobo** [comment]
> Just be clear, this is fixed recurrence of let's say `r`, evaluated at `r`?

@JonasGeiping  Sorry, I forgot to clear this, what I mentioned of recurrent `r` is training at fixed `r`

**Comment by JonasGeiping** [comment]
Yes, I understand - what I'm trying to say is that training at `r` and evaluating at `r` is expected to be better than training with randomized unrolling (by definition almost). For us, the gain was marginal (small), but a gain (so it was positive) when using fixed `r`, which is exactly what you are also observing.

**Comment by shuxiaobo** [comment]
> Yes, I understand - what I'm trying to say is that training at `r` and evaluating at `r` is expected to be better than training with randomized unrolling (by definition almost). For us, the gain was marginal (small), but a gain (so it was positive) when using fixed `r`, which is exactly what you are also observing.
@JonasGeiping 
and if i want to monitoring the val loss when training the model using random recurrent `r`, the val recurrent `r` should be fix or max recurrent times?
